[Apzure] iP

Apzure opened the pull request:
### Duke2

> "Your mind is for having ideas, not holding them." – David Allen

DukePro frees your mind of having to remember things you need to do.  It is 

- text-based
- *easy* to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,

1. Download it from [here](https://github.com/Apzure/ip/releases/tag/A-Jar).
2. Double-click it.
3. Add your tasks.
4. Let it manage your tasks for you 😸 
5. And it is **FREE!**

## Features:
- [x] Managing tasks
- [ ]  Managing deadlines (coming soon)
- [ ]  Reminders (coming soon)


`def function()`

If you are a Java programmer, you can use it to practice Java too. Here's the main method:

```java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

Commit:
Level-0

Commit:
Level-1

Commit:
Level-2 and Level-3

Commit:
Level-4

Commit:
Level-5

Commit:
Level-6

Commit:
More Error-Handling

Commit:
More Error Handling

Commit:
A-TextUiTesting

Commit:
Fix and better Error Handling

Commit:
Error handling for event command arguments

Commit:
A-Enums

Commit:
Custom Message if you have no tasks when doing list command

Commit:
Updated testing accordingly

Commit:
Error Handling if the idx for commands like mark, unmark is out of range

Commit:
Initial File and Folder Creation for Data Storage

Commit:
More Testing

Commit:
Segemation of Tasklist into a new class, away from Bird

Commit:
Functionality to Store TaskList whenever it is updated

Commit:
Storage fully functional

Commit:
Bugfix for error handling of invalid commands

Commit:
Cleaning up import

Commit:
Updated test

Commit:
Removed Debugging print statement

Commit:
Removed

Commit:
Merge branch 'branch-Level-7'

Commit:
LocalDate Format implemented

Commit:
Merge branch 'branch-Level-8'

Commit:
Initial Creation of UI

Commit:
Updated testing wrt to dates

Commit:
Ui class

Commit:
Put all classes in duke package

Commit:
Renamed package from Duke to duke

Commit:
Merge branch 'add-gradle-support'

Commit:
Gradle support

Commit:
JUnit Tests for CommandHandler markTask method and Storage init method.

Commit:
JavaDoc Comments

Commit:
Coding Standards

Commit:
Fix Spelling Error

Commit:
Remove Unused Import and variable

Commit:
Find Implemented

Commit:
Merge branch 'branch-A-JavaDoc'

Commit:
Merge branch 'branch-A-CodingStandard'

Commit:
Adding data to .gitignore

Commit:
Merge branch 'branch-Level-9'

Commit:
Add Basic GUI outline

Created a Basic GUI to serve as a test by following tutorial

To be improved upon and integrated with chatbot in future commits

Commit:
Add User Interaction Outline Standalone

User Interaction outline added by following tutorial provided

Serves as a trial run, does not implement the functionalities of
the chatbot

Commit:
Integrate GUI with Duke Functionality

GUI works with Duke and has replaced the Ui

GUI formatting still needs work

Commit:
Add Integration with FXML

GUI now uses FXML and has better separation of logic

Does not give greet message on startup yet

Commit:
Add welcome message on startup

Commit:
Add Basic Styling for GUI

Commit:
Merge branch 'branch-Level-10'

JerryO3 commented on src/main/java/Duke/InvalidDateFormat.java:
```diff
@@ -0,0 +1,7 @@
+package duke;
+
+public class InvalidDateFormat extends DukeException {
```
Should this exception name follow the existing naming pattern you've established i.e. [description]Exception?

```suggestion
public class InvalidDateFormatException extends DukeException {
```

Apzure replied:
Thanks for all the comments! Your suggestions make sense, and I'll be sure to improve with those in mind!

JerryO3 commented on src/main/java/Duke/IndexOutOfRange.java:
```diff
@@ -0,0 +1,7 @@
+package duke;
+
+public class IndexOutOfRange extends DukeException {
```
Should this exception name follow the existing naming pattern you've established i.e. [description]Exception?

```suggestion
public class IndexOutOfRangeException extends DukeException {
```

JerryO3 commented on src/main/java/Duke/Bird.java:
```diff
@@ -0,0 +1,66 @@
+package duke;
```
I like how you've made your Duke equivalent class readable using strong OOP principles. Was pleasantly surprised to find that the main application driver is so concise!

JerryO3 commented on src/main/java/Duke/CommandHandler.java:
```diff
@@ -0,0 +1,158 @@
+        String[] words = userInput.split("\\s+");
+        Command command;
+        try {
+            command = Command.valueOf(words[0].toUpperCase());
```
Really like this implementation of command parsing! Noted that this allows commands to be case-insensitive 👍

JerryO3 commented on src/main/java/Duke/Storage.java:
```diff
@@ -0,0 +1,147 @@
+        }
+    }
+}
+
```
Would it be better if you provided more documentation on how the store and decode works? I only realised that you were storing the task as a string directly in the toString() format of each Task and then using decode() to parse this stored string back into a representative ItemList object after cross referencing both your store() and decode() methods. 

There may be other ways of doing this which may warrant your documentation on how you've implemented storage functionality. Otherwise, its a pretty neat implementation of storing Tasks 👍

JerryO3 reviewed (commented):
LGTM, some nits include your naming conventions regarding exceptions. Otherwise your code largely adheres to the coding standards! Really liked your implementation of command parsing as well :)

Commit:
Add Assertions

Added several assertions for early checking of errors

For example
*checking that the output of duke is never null
*ensuring switch block does not fall into default for execute
command and decode methods

Commit:
Merge branch 'branch-A-Assertions'

Commit:
Rename exceptions so that they end with "Exception"

Most custom-made exceptions already ended with exception
etc InvalidEventFormatException

There were a few outliers to this rule, so they have been renamed
to follow this standard for clarity purposes

Commit:
Segment executeCommand into multiple methods

Commit:
Amend outdated comment and variable

Commit:
Handle duke's messages being cut off

Commit:
Rename of method name

Commit:
Delegate the encode and decode of tasks to themselves

Initially all this processing was done by storage.java

Now, it is given out to the task classes themselves

This is more in alignment with OOP principles

It also makes storage.java more readable and understandable

Commit:
Rename variables for readability

Commit:
Update .gitignore to remove outdated text-ui

Commit:
Merge branch 'branch-A-CodeQuality'

Commit:
Add Sorting Functionality

It sorts by end date first

ToDo are always given less priority to tasks with dates

ToDo objects are sorted among one another based on their name

Apzure committed:
Merge pull request #1 from Apzure/branch-BCD-Extension

Add Sorting Functionality

Commit:
Remove text-ui-test

Obsolete now that GUI is being used

Commit:
Add UI image and update ReadMe with UG

Commit:
Update UG

Apzure committed:
Update README.md

Apzure committed:
Update README.md

Apzure committed:
Update README.md

Commit:
Add JavaDoc comments for Exceptions

Commit:
Change Image of Bird

Commit:
Update Ui picture
